### PR TITLE
refactor(storage): OpenDAL list object should be compatible with s3

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
OpenDAL provide `scan` recently, which will list all objects instead of all dirs, we should use `scan` to be compatible with s3.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
~~- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).~~
~~- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)~~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [x] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/workspace-hack/Cargo.toml [hunk 1]
```diff
@@ -172,7 +172,8 @@ serde_json = { version = "1", features = ["alloc"] }
 serde_with = { version = "2", features = ["json"] }
 smallvec = { version = "1", default-features = false, features = ["serde"] }
 subtle = { version = "2" }
-syn = { version = "1", features = ["extra-traits", "full", "visit", "visit-mut"] }
+syn-dff4ba8e3ae991db = { package = "syn", version = "1", features = ["extra-traits", "full", "visit", "visit-mut"] }
+syn-f595c2ba2a3f28df = { package = "syn", version = "2", features = ["full", "visit-mut"] }
```
Comment (Gun9niR) [inline src/workspace-hack/Cargo.toml:176]: Why is this needed? Seems syn is not used in this PR.

Comment (wcy-fdu) [reply src/workspace-hack/Cargo.toml:176]: May related to OpenDAL version `0.33.1`? I just bump version.

**Review by Gun9niR — APPROVED** [review]
Rest LGTM. BTW, S3 supports listing paths not ending with `/`. While OpenDAL has `list_with` to list dirs by prefix, `scan_with` has yet to be implemented. If we want to align OpenDAL's `list` semantic to S3, we can migrate to `scan_with` after it's implemented.

File: src/workspace-hack/Cargo.toml (hunk 1 above)
Comment (wcy-fdu) [reply src/workspace-hack/Cargo.toml:176]: Can we ban `typo checker / Spell Check with Typos (pull_request) ` for .toml?
